### PR TITLE
Use ident.span instead of def_span in dead-code pass

### LINKED ISSUE #58729: Unused function warnings should return a span of only the function name
Currently a warning for an unused function returns a span of the entire function.  As a consequence, RLS based IDE environments may then put green squiggles on the entire function. This makes it very annoying to work on a function, if you have not yet written the function that uses it. 

There may be other unused warnings for structs/traits that could be similarly adjusted.

This issue has been assigned to @Quantumplation via [this comment](https://github.com/rust-lang/rust/issues/58729#issuecomment-545980897).

### PR DESCRIPTION
**Pull request by Quantumplation**
Hello! First time contributor! :)

This should fix #58729.

According to @estebank in the duplicate #63064, def_span scans forward on the line until it finds a {,
and if it can't find one, falls back to the span for the whole item. This
was apparently written before the identifier span was explicitly tracked on
each node.

This means that if an unused function signature spans multiple lines, the
entire function (potentially hundreds of lines) gets flagged as dead code.
This could, for example, cause IDEs to add error squiggly's to the whole
function.

By using the span from the ident instead, we narrow the scope of this in
most cases. In a wider sense, it's probably safe to use ident.span
instead of def_span in most locations throughout the whole code base,
but since this is my first contribution, I kept it small.

Some interesting points that came up while I was working on this:
- I reorganized the tests a bit to bring some of the dead code ones all
into the same location
- A few tests were for things unrelated to dead code (like the
path-lookahead for parens), so I added #![allow(dead_code)] and
cleaned up the stderr file to reduce noise in the future
- The same fix doesn't apply to const and static declarations. I tried
adding these cases to the match expression, but that created a much
wider change to tests and error messages, so I left it off until I
could get some code review to validate the approach.

### CODE REVIEW

**Comment by Quantumplation** [comment]
Hmm, more failures... I'm not sure if these tests ran when I ran tests locally...  Fixing now!

**Comment by Quantumplation** [comment]
I got a seemingly unrelated test failure when running locally.  given the error message, it sounds like it might just be a test that's not intended to run on WSL? I'm not sure...

```
failures:

---- net::tcp::tests::double_bind stdout ----
thread '<unnamed>' panicked at 'This system (perhaps due to options set by TcpListener::bind) permits double binding: TcpListener { addr: V4(127.0.0.1:19613), fd: 8 } and TcpListener { addr: V4(127.0.0.1:19613), fd: 15 }', src/libstd/net/tcp.rs:1257:34


failures:
    net::tcp::tests::double_bind

test result: FAILED. 762 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
```

This gets thrown here: https://github.com/rust-lang/rust/blob/master/src/libstd/net/tcp.rs#L1257

I'm going to push my changes and hope the build passes when running on the linux build server :sweat_smile:

**Review by davidtwco — APPROVED** [review]
This looks good to me. If you don't mind, I'd prefer the last three commits (84ecd026184e4979beb43d1aaf1a4e83c44bacc1, 0e00a56746d868c81bad604bca107364ef775b12 and 9b5ca693fdf731beaf88332b813f90bf3f0031ca) be squashed into 1277650b5fed98c19ee56c2bcf68fdc9bacaf7b5.

**Comment by Quantumplation** [comment]
Sure thing!

**Comment by davidtwco** [comment]
Thanks @Quantumplation!

@bors r+

**Comment by Quantumplation** [comment]
At @estebank's request, rebase this onto master and re-blessing the tests

**Comment by Quantumplation** [comment]
I wasn't able to run tests to completion (same bind_address issue I mentioned above), but it got pretty far and no tests were updated.  I'm pushing the rebase regardless.

**Comment by Centril** [comment]
Possibly caused failure in https://github.com/rust-lang/rust/pull/65854#issuecomment-546648533, @bors rollup=never

**Comment by Quantumplation** [comment]
@Centril what is your recommended path to resolution on this?

**Comment by Centril** [comment]
@Quantumplation Check if something in your PR could have caused this, otherwise CI will tell us eventually.

**Comment by Quantumplation** [comment]
@Centril Hmm..  Looks like it passed without my changes, but I'm pretty baffled as to why my changes would cause failures.  It doesn't look like an .stderr that I forgot to update, and all I did was change how the error message was printing.  I'll try to dig deeper later today when I'm back at the computer I use for rust dev.

**Comment by davidtwco** [comment]
> 
> 
> @Centril Hmm.. Looks like it passed without my changes, but I'm pretty baffled as to why my changes would cause failures. It doesn't look like an .stderr that I forgot to update, and all I did was change how the error message was printing. I'll try to dig deeper later today when I'm back at the computer I use for rust dev.

@Quantumplation Rust's CI merges with master before it runs any tests. If a new test was added in master that your patch would have affected, then it won't have an updated stderr (because your patch doesn't know about it). When @Centril attempted a rollup (combining lots of patches to merge all at once) yesterday, the failure looked like it might have been because of what I've just described happening with your patch.

That's probably why @estebank requested you rebase and re-bless. Since you did that, so long as more new tests haven't been added that would fail in the same way, your patch should be good to merge, which is why I re-approved it. But given that your patch seems susceptible to this sort of failure (as evidenced by it having happened), @Centril marked it as `rollup=never`, which means that it won't be combined with other small patches and landed all at once (since if it *did* cause a failure, it slows down other patches landing too).

tl;dr you probably don't need to do anything right now, by rebasing and re-blessing yesterday you've probably fixed the problem.

**Comment by Quantumplation** [comment]
@davidtwco Ah, right, that makes sense.  For some reason I thought this failed even after I rebased, but it didn't make it's way into a roll up, so everything makes sense again heh.

File: src/test/ui/span/macro-span-replacement.stderr [hunk 1]
```diff
@@ -1,11 +1,8 @@
 warning: struct is never constructed: `S`
-  --> $DIR/macro-span-replacement.rs:7:14
+  --> $DIR/macro-span-replacement.rs:7:12
    |
 LL |         $b $a;
-   |              ^
-...
-LL |     m!(S struct);
-   |     ------------- in this macro invocation
+   |            ^^
```
Comment (bjorn3) [inline src/test/ui/span/macro-span-replacement.stderr]: This is a error quality degradation.

Comment (Quantumplation) [reply src/test/ui/span/macro-span-replacement.stderr]: Mmm. I was looking at this before I submit the PR, and was a bit confused as to why my changes would impact this.

Now that I've had some time to get more familiar with the rust code base, I see that here:

https://github.com/rust-lang/rust/blob/master/src/librustc_errors/emitter.rs#L354-L366

We check if the span is inside of a macro.  Previously, on this code, `def_span(item.span)` would just return `item.span` (because it didn't find a {), which *was* considered inside of a macro (not sure why?), whereas ident.span is not conisdered inside a macro. 

Not sure how to resolve this, tbh.  Open to advice from people with more experience with the code base!

**Comment by Quantumplation** [comment]
After some digging, I *think* the test failure and the error message regression reported by @bjorn3 are caused by the same thing, i.e. a change in the macro expansion.

Specifically, see this line: https://github.com/rust-lang/rust/blob/master/src/librustc/lint/mod.rs#L681

Here, it uses *the reported span alone* to detect if the error is in an external macro.  Previously, (I guess, I still don't understand well enough to say definitively), `def_span(...)` would return a span that other code would consider to be part of a macro.  This would do two things:

 - trigger the "in this macro invocation" helper above
 - cause the error from the `query`, `Keywords` and `Symbol` unused symbols to be canceled.

One thing that's unclear to me is why `query`, `Keywords`, and `Symbol` are considered dead code in the first place: they're used elsewhere in the file (example: https://github.com/rust-lang/rust/blob/master/src/librustc_macros/src/query.rs#L153)

Until someone can weigh in on how to address this more thoroughly, I'll just revert to the old def_span call if `item.span` is in a macro.  Exactly how to do that is TBD, while I poke around the code, but I'll be looking at these two places for hints, since this is how the two cases above detect macros:

https://github.com/rust-lang/rust/blob/master/src/librustc/lint/mod.rs#L898
https://github.com/rust-lang/rust/blob/master/src/libsyntax_pos/lib.rs#L133

File: src/test/ui/span/macro-span-replacement.stderr (hunk 1 above)
Comment (estebank) [reply src/test/ui/span/macro-span-replacement.stderr]: I believe that the `item.span` was marked with `with_ctxt`, whereas this new span wasn't, because in the current model it is true that this span is fully enclosed by the macro and has no other context. I would merge this PR as is with a follow up ticket filed to address this case. It might be a good idea to mark all the spans that involve a macro expansion site with a context so that these type of errors highlight the caller as well. That change would likely cause multiple other changes in other errors.

**Comment by estebank** [comment]
> but I'll be looking at these two places for hints, since this is how the two cases above detect macros:

I believe the later `libsyntax_pos::is_macros` would work for you.

File: src/test/ui/span/macro-span-replacement.stderr (hunk 1 above)
Comment (Quantumplation) [reply src/test/ui/span/macro-span-replacement.stderr]: @estebank roger, opened a different issue for it.

**Comment by Quantumplation** [comment]
Sorry, had a bit of a mixup with my commits, forgot to configure user.name and user.email on a new setup so I had to force push :)

File: src/librustc_passes/dead.rs [hunk 2]
```diff
@@ -569,16 +569,26 @@ impl Visitor<'tcx> for DeadVisitor<'tcx> {
...
                 hir::ItemKind::Union(..) |
                 hir::ItemKind::Trait(..) |
-                hir::ItemKind::Impl(..) => self.tcx.sess.source_map().def_span(item.span),
+                hir::ItemKind::Impl(..) => {
+                    // FIXME(66095): Because item.span is annotated with things
+                    // like a macro_backtrace, and ident.span isn't, we use the
+                    // def_span method if it's part of a macro invocation
+                    // We should probably annotate ident.span with the macro
+                    // context, but that's a larger change.
+                    if item.span.macro_backtrace().len() == 0 {
```
Comment (estebank) [inline src/librustc_passes/dead.rs]: would `item.span.source_callee().is_none()` work here instead of building the macro backtrace?

**Comment by estebank** [comment]
The hacky solution is ok by me.

File: src/librustc_passes/dead.rs (hunk 2 above)
Comment (Quantumplation) [reply src/librustc_passes/dead.rs]: I'll give it a try.  Can you explain what source_callee is doing for my own education?

Comment (estebank) [reply src/librustc_passes/dead.rs]: If a span is inside a macro, they have expansion data, including among other things where the span for the macro call is coming from, which `source_callee` returns. If a span has no expansion data, it means that it isn't coming from a macro context.
